test/php/gutenberg-coding-standards/Gutenberg/Sniffs/Commenting/FunctionCommentSinceTagSniff.php

```diff
@@ -48,90 +91,444 @@ public function process( File $phpcsFile, $stackPtr ) {
 			return;
 		}
 
-		$tokens        = $phpcsFile->getTokens();
-		$function_name = $phpcsFile->getDeclarationName( $stackPtr );
+		$tokens = $phpcsFile->getTokens();
+		$token  = $tokens[ $stackPtr ];
+
+		if ( 'T_FUNCTION' === $token['type'] ) {
+			$this->process_function_token( $phpcsFile, $stackPtr );
+			return;
+		}
+
+		if ( isset( static::$oo_tokens[ $token['code'] ] ) ) {
+			$this->process_oo_token( $phpcsFile, $stackPtr );
+			return;
+		}
 
-		$wrapping_tokens_to_check = array(
-			T_CLASS,
-			T_INTERFACE,
-			T_TRAIT,
+		if ( 'T_STRING' === $token['type'] ) {
+			$this->process_hook( $phpcsFile, $stackPtr );
+			return;
+		}
+
+		if ( 'T_VARIABLE' === $token['type'] && Scopes::isOOProperty( $phpcsFile, $stackPtr ) ) {
+			$this->process_property_token( $phpcsFile, $stackPtr );
+		}
+	}
+
+	/**
+	 * Processes a token representing a function call that invokes a WordPress hook,
+	 * checking for a missing `@since` tag in its docblock.
+	 *
+	 * @param File $phpcs_file    The file being scanned.
+	 * @param int  $stack_pointer The position of the hook token in the stack.
+	 */
+	protected function process_hook( File $phpcs_file, $stack_pointer ) {
+		$tokens = $phpcs_file->getTokens();
+
+		// The content of the current token.
+		$hook_function = $tokens[ $stack_pointer ]['content'];
+
+		$hook_invocation_functions = array(
+			'do_action',
+			'do_action_ref_array',
+			'apply_filters',
+			'apply_filters_ref_array',
 		);
 
-		foreach ( $wrapping_tokens_to_check as $wrapping_token_to_check ) {
-			if ( false !== $phpcsFile->getCondition( $stackPtr, $wrapping_token_to_check, false ) ) {
-				// This sniff only processes functions, not class methods.
+		// Check if the current token content is one of the filter functions.
+		if ( ! in_array( $hook_function, $hook_invocation_functions, true ) ) {
+			// Not a hook.
+			return;
+		}
+
+		$missing_since_tag_error_message = sprintf(
+			'@since tag is missing for the "%s()" hook function.',
+			$hook_function
+		);
+
+		$violation_code = 'MissingHookSinceTag';
+
+		$docblock = static::find_hook_docblock( $phpcs_file, $stack_pointer );
+		if ( false === $docblock ) {
+			$phpcs_file->addError( $missing_since_tag_error_message, $stack_pointer, $violation_code );
+			return;
+		}
+
+		list( $doc_block_start_token, $doc_block_end_token ) = $docblock;
+
+		$version_token = static::parse_since_tags( $phpcs_file, $doc_block_start_token, $doc_block_end_token );
+		if ( false === $version_token ) {
```

Comment:
Is there a scenario I'm missing where `$version_token` is `false`?

If there is, maybe calling `FunctionCommentSinceTagSniff::parse_since_tags()` and `GetTokensAsString::compact()` twice inside this method is unnecessary? I haven't spent too much time checking the code, but it is not immediately clear to me the difference between the two code blocks where those two functions are called.

Reply:
You are right. This block of code isn't needed. Removed in 7a0bd41b44823b4834e91bbefe2568d3bf8e26c6